Bluetooth: host: Periodic scanner drop incomplete data

Pieloaf opened the pull request:
This changes the periodic scanner to drop periodic adv reports with incomplete data. This avoids incorrect data being sent to application in the case where the periodic adv data is not successfully received by the scanner.

Fixes #54072

Pieloaf committed:
bluetooth: host: refactor bt_hci_le_per_adv_report data reassembly

Refactoring the periodic advertisment report receive function
to make the handling of data reassembly more readable.

Signed-off-by: Pierce Lowe <pierce.lowe@nordicsemi.no>

jori-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -805,6 +805,11 @@ void bt_hci_le_per_adv_report(struct net_buf *buf)
 			bt_hci_le_per_adv_report_recv(per_adv_sync, &buf->b, &info);
 		} else {
+			if (evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_INCOMPLETE) {
+				per_adv_sync->report_truncated = true;
```
can you add a `LOG_DBG` as is done a L815?

Pieloaf commented:
most recent force push was a rebase to align with main

jori-nordic reviewed (changes requested):
I forgot on the first pass:
Commit title should be: `Bluetooth: host: drop incomplete data in periodic scanner`

Thalley commented on subsys/bluetooth/host/scan.c:
```diff
@@ -805,6 +805,14 @@ void bt_hci_le_per_adv_report(struct net_buf *buf)
 			 */
 			bt_hci_le_per_adv_report_recv(per_adv_sync, &buf->b, &info);
 		} else {
+			if (evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_INCOMPLETE) {
+				LOG_WRN("Received incomplete advertising data. "
+					"Advertising report dropped.");
+
+				per_adv_sync->report_truncated = true;
+				net_buf_simple_reset(&per_adv_sync->reassembly);
+				return;
+			}
```
Given that we `return` here, shouldn't this block be put into an `else if`, instead of the current approach?

Pieloaf replied:
Yeah I wondered that as well, but did it this way since it was done the same way in [the lines below it ](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/scan.c#L808-L817). But can move it to its own if else if that is preferred.

Thalley replied:
I would prefer an `else if` here. 

It is a bit tricky given the existing structure, but ideally it should have been
```c
if (evt->status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE) {
    /* do stuff */
} else if (evt->status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_INCOMPLETE) {
    /* do stuff */
} else if (evt->status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE) {
   /* do stuff */ 
}
```

Pieloaf replied:
I've restructured the function a bit to follow the structure you suggested. Let me know if it looks ok.

Thalley commented on subsys/bluetooth/host/scan.c:
```diff
@@ -798,29 +798,40 @@ void bt_hci_le_per_adv_report(struct net_buf *buf)
+		}
+
+		if (evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE) {
+			if(per_adv_sync->reassembly.len == 0) {
```
```suggestion
			if (per_adv_sync->reassembly.len == 0) {
```

Thalley commented on subsys/bluetooth/host/scan.c:
```diff
@@ -798,29 +798,40 @@ void bt_hci_le_per_adv_report(struct net_buf *buf)
+			per_adv_sync->report_truncated = true;
+			net_buf_simple_reset(&per_adv_sync->reassembly);
+
+		} else if { evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL
+			net_buf_simple_add_mem(&per_adv_sync->reassembly, buf->data, evt->length);
```
```suggestion
		} else if ( evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL) {
			net_buf_simple_add_mem(&per_adv_sync->reassembly, buf->data, evt->length);
```

Thalley reviewed (changes requested):
I think this new structure is easier to read - Thanks! 

A few bugs in the code though

Pieloaf commented:
> I think this new structure is easier to read - Thanks!
> 
> A few bugs in the code though

Yeah sorry I tested it in a different environment locally and fixed those bugs there but missed them when I pushed from the environment Im making the PR from, working on it now.

Thalley commented:
Hmm, unless I'm mistaken, it actually doesn't look like we are setting `CONFIG_BT_PER_ADV_SYNC_BUF_SIZE > 0` anywhere or testing that code at all. 

We should add at test for that before we merge this PR

Pieloaf commented:
Oh hmmm, good catch. The environment I'm testing it in is using that code since I'm getting the warning messages I've added when packets are dropped, but yes it should be tested properly in CI

Thalley commented:
> Oh hmmm, good catch. The environment I'm testing it in is using that code since I'm getting the warning messages I've added when packets are dropped, but yes it should be tested properly in CI

Could I bother you to add this as a babblesim test or something? At the very least we should build the code in CI :D

Pieloaf commented:
> > Oh hmmm, good catch. The environment I'm testing it in is using that code since I'm getting the warning messages I've added when packets are dropped, but yes it should be tested properly in CI
> 
> Could I bother you to add this as a babblesim test or something? At the very least we should build the code in CI :D

I've not used babblesim before but I can try get a simple test setup with it today. Should I make a second PR for it or is ok as part of this one?

Thalley commented:
> > > Oh hmmm, good catch. The environment I'm testing it in is using that code since I'm getting the warning messages I've added when packets are dropped, but yes it should be tested properly in CI
> > 
> > 
> > Could I bother you to add this as a babblesim test or something? At the very least we should build the code in CI :D
> 
> I've not used babblesim before but I can try get a simple test setup with it today. Should I make a second PR for it or is ok as part of this one?

The test should be added in this PR IMO

Pieloaf commented:
@Thalley I wrote a babble sim test at the end of last week but I'm getting an I/O error when trying to set the advertising data. I referenced other bsim tests for it and I'm not sure where I'm going wrong. If I push the changes here would you mind taking a look if you have the time?

jori-nordic commented:
> @Thalley I wrote a babble sim test at the end of last week but I'm getting an I/O error when trying to set the advertising data. I referenced other bsim tests for it and I'm not sure where I'm going wrong. If I push the changes here would you mind taking a look if you have the time?

I can help you out, contact me on teams

Pieloaf committed:
tests: bluetooth: added bsim test for longer per adv data

Added a bsim_bt host test to test the handling of periodic adv
report data received across multiple packets

Signed-off-by: Pierce Lowe <pierce.lowe@nordicsemi.no>

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_advertiser.c:
```diff
@@ -21,6 +21,29 @@ extern enum bst_result_t bst_result;
 static struct bt_conn *g_conn;
 
+static uint8_t mfg_data[] = {
+	0xFF, 0xFF, 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B, 0x0C,
+	0x0D, 0x0E, 0x0F, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1A, 0x1B,
+	0x1C, 0x1D, 0x1E, 0x1F, 0x20, 0x21, 0x22, 0x23, 0x24, 0x25, 0x26, 0x27, 0x28, 0x29, 0x2A,
+	0x2B, 0x2C, 0x2D, 0x2E, 0x2F, 0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39,
+	0x3A, 0x3B, 0x3C, 0x3D, 0x3E, 0x3F, 0x40, 0x41, 0x42, 0x43, 0x44, 0x45, 0x46, 0x47, 0x48,
+	0x49, 0x4A, 0x4B, 0x4C, 0x4D, 0x4E, 0x4F, 0x50, 0x51, 0x52, 0x53, 0x54, 0x55, 0x56, 0x57,
+	0x58, 0x59, 0x5A, 0x5B, 0x5C, 0x5D, 0x5E, 0x5F, 0x60, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66,
+	0x67, 0x68, 0x69, 0x6A, 0x6B, 0x6C, 0x6D, 0x6E, 0x6F, 0x70, 0x71, 0x72, 0x73, 0x74, 0x75,
+	0x76, 0x77, 0x78, 0x79, 0x7A, 0x7B, 0x7C, 0x7D, 0x7E, 0x7F, 0x80, 0x81, 0x82, 0x83, 0x84,
+	0x85, 0x86, 0x87, 0x88, 0x89, 0x8A, 0x8B, 0x8C, 0x8D, 0x8E, 0x8F, 0x90, 0x91, 0x92, 0x93,
+	0x94, 0x95, 0x96, 0x97, 0x98, 0x99, 0x9A, 0x9B, 0x9C, 0x9D, 0x9E, 0x9F, 0xA0, 0xA1, 0xA2,
+	0xA3, 0xA4, 0xA5, 0xA6, 0xA7, 0xA8, 0xA9, 0xAA, 0xAB, 0xAC, 0xAD, 0xAE, 0xAF, 0xB0, 0xB1,
+	0xB2, 0xB3, 0xB4, 0xB5, 0xB6, 0xB7, 0xB8, 0xB9, 0xBA, 0xBB, 0xBC, 0xBD, 0xBE, 0xBF, 0xC0,
+	0xC1, 0xC2, 0xC3, 0xC4, 0xC5, 0xC6, 0xC7, 0xC8, 0xC9, 0xCA, 0xCB, 0xCC, 0xCD, 0xCE, 0xCF,
+	0xD0, 0xD1, 0xD2, 0xD3, 0xD4, 0xD5, 0xD6, 0xD7, 0xD8, 0xD9, 0xDA, 0xDB, 0xDC, 0xDD, 0xDE,
+	0xDF, 0xE0, 0xE1, 0xE2, 0xE3, 0xE4, 0xE5, 0xE6, 0xE7, 0xE8, 0xE9, 0xEA, 0xEB, 0xEC, 0xED,
+	0xEE, 0xEF, 0xF0, 0xF1, 0xF2, 0xF3, 0xF4, 0xF5, 0xF6, 0xF7, 0xF8, 0xF9, 0xFA, 0xFB};
```
I'm curious about this array: Is the data specifically defined? 

Just wondering why it's done this way, instead of "just" filling it with an array like
```c
for (size_t i = 0U; i < ARRAY_SIZE(mfg_data); i++) {
    mfg_data[i] = i;
}
```

Pieloaf replied:
I initially had it like this with 
```C
for (size_t i = 0U; i < ARRAY_SIZE(mfg_data); i++) {
    mfg_data[i] = i;
}

// manufacturer data
mfg_data[0] = 0xFF;
mgf_data[1] = 0xFF; 
```
But looking at other cases which set mfg_data, I couldn't find a single case where the data wasn't specifically defined so changed it to match. But can change it back.

Thalley replied:
Ah, so one of the `0xff` is for the type? 

If using `BT_DATA` or `BT_DATA_BYTES` the type and length will be set by the macros, and don't need to be specified in the data itself.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_advertiser.c:
```diff
@@ -21,6 +21,29 @@ extern enum bst_result_t bst_result;
+	0xEE, 0xEF, 0xF0, 0xF1, 0xF2, 0xF3, 0xF4, 0xF5, 0xF6, 0xF7, 0xF8, 0xF9, 0xFA, 0xFB};
+
+
+static const struct bt_data ad[] = {BT_DATA(BT_DATA_MANUFACTURER_DATA, mfg_data, 254),
+				    BT_DATA(BT_DATA_MANUFACTURER_DATA, mfg_data, 254)};
```
```suggestion
static const struct bt_data ad[] = {BT_DATA_BYTES(BT_DATA_MANUFACTURER_DATA, mfg_data),
				    BT_DATA_BYTES(BT_DATA_MANUFACTURER_DATA, mfg_data)};
```

Thalley replied:
Should we also consider trying with the maximum size of 1650?

Pieloaf replied:
I did try with 1650, but I could only send up to 1004 bytes. After that the application wasn't receiving the data, it may have been due to the periodic adv parameters im not sure but just reduced as it wasn't a test to ensure the advertiser can advertise maximum per adv data but that the scanner reassembles the split data correctly

Pieloaf replied:
I was getting build errors with BT_DATA_BYTES as it was expecting a char/uint8_t not a uint8_t *
Looking at other instances where BT_DATA_BYTES is used the data bytes are inline in the function and not a pointer to an array of the data. So have left it as BT_DATA

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_advertiser.c:
```diff
@@ -145,6 +168,20 @@ static void start_per_adv_set(struct bt_le_ext_adv *adv)
+		printk("Failed to set periodic advertising data: %d\n",
+		       err);
+		return;
+	}
+	printk("done.\n");
```
```suggestion
	}
	
	printk("done.\n");
```

Pieloaf replied:
The other functions in the test don't space the printk("done.\n") from the brace which is why i hadn't done it here
[1](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_advertiser.c#LL106), [2](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_advertiser.c#L119), [3](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_advertiser.c#L132)...

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_advertiser.c:
```diff
@@ -276,6 +313,31 @@ static void main_per_adv_conn_privacy_advertiser(void)
 
+static void main_per_adv_long_data_advertiser(void)
+{
+#if (CONFIG_BT_PER_ADV_SYNC_BUF_SIZE > 0)
```
`CONFIG_BT_PER_ADV_SYNC_BUF_SIZE` only applies to the sync receiver, right? What's the argument for guarding the advertiser side?

Pieloaf replied:
Could use either of these for advertiser:
`CONFIG_BT_CTLR_ADV_DATA_LEN_MAX`
`CONFIG_BT_CTLR_ADV_DATA_CHAIN`
which I did initially use the ADV_DATA_LEN_MAX > 31 but set them both to the SYNC_BUF_SIZE, since it was only used in this test and they both used the same conf file. But for readability I can change it back

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_syncer.c:
```diff
@@ -22,12 +22,35 @@ extern enum bst_result_t bst_result;
 static struct bt_conn *g_conn;
 static bt_addr_le_t per_addr;
 static uint8_t per_sid;
+static uint8_t mfg_data[] = {
+	0xFF, 0xFF, 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B, 0x0C,
+	0x0D, 0x0E, 0x0F, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1A, 0x1B,
+	0x1C, 0x1D, 0x1E, 0x1F, 0x20, 0x21, 0x22, 0x23, 0x24, 0x25, 0x26, 0x27, 0x28, 0x29, 0x2A,
+	0x2B, 0x2C, 0x2D, 0x2E, 0x2F, 0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39,
+	0x3A, 0x3B, 0x3C, 0x3D, 0x3E, 0x3F, 0x40, 0x41, 0x42, 0x43, 0x44, 0x45, 0x46, 0x47, 0x48,
+	0x49, 0x4A, 0x4B, 0x4C, 0x4D, 0x4E, 0x4F, 0x50, 0x51, 0x52, 0x53, 0x54, 0x55, 0x56, 0x57,
+	0x58, 0x59, 0x5A, 0x5B, 0x5C, 0x5D, 0x5E, 0x5F, 0x60, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66,
+	0x67, 0x68, 0x69, 0x6A, 0x6B, 0x6C, 0x6D, 0x6E, 0x6F, 0x70, 0x71, 0x72, 0x73, 0x74, 0x75,
+	0x76, 0x77, 0x78, 0x79, 0x7A, 0x7B, 0x7C, 0x7D, 0x7E, 0x7F, 0x80, 0x81, 0x82, 0x83, 0x84,
+	0x85, 0x86, 0x87, 0x88, 0x89, 0x8A, 0x8B, 0x8C, 0x8D, 0x8E, 0x8F, 0x90, 0x91, 0x92, 0x93,
+	0x94, 0x95, 0x96, 0x97, 0x98, 0x99, 0x9A, 0x9B, 0x9C, 0x9D, 0x9E, 0x9F, 0xA0, 0xA1, 0xA2,
+	0xA3, 0xA4, 0xA5, 0xA6, 0xA7, 0xA8, 0xA9, 0xAA, 0xAB, 0xAC, 0xAD, 0xAE, 0xAF, 0xB0, 0xB1,
+	0xB2, 0xB3, 0xB4, 0xB5, 0xB6, 0xB7, 0xB8, 0xB9, 0xBA, 0xBB, 0xBC, 0xBD, 0xBE, 0xBF, 0xC0,
+	0xC1, 0xC2, 0xC3, 0xC4, 0xC5, 0xC6, 0xC7, 0xC8, 0xC9, 0xCA, 0xCB, 0xCC, 0xCD, 0xCE, 0xCF,
+	0xD0, 0xD1, 0xD2, 0xD3, 0xD4, 0xD5, 0xD6, 0xD7, 0xD8, 0xD9, 0xDA, 0xDB, 0xDC, 0xDD, 0xDE,
+	0xDF, 0xE0, 0xE1, 0xE2, 0xE3, 0xE4, 0xE5, 0xE6, 0xE7, 0xE8, 0xE9, 0xEA, 0xEB, 0xEC, 0xED,
+	0xEE, 0xEF, 0xF0, 0xF1, 0xF2, 0xF3, 0xF4, 0xF5, 0xF6, 0xF7, 0xF8, 0xF9, 0xFA, 0xFB};
```
Instead of defining this both places, I'd suggest to just put it in the `common.h` file

Pieloaf replied:
This was my very first approach but then ran into issue with compiling it getting errors that it was "defined more than once" but both times it pointed to common.h so I had moved it to the individual files, but I'll see if I can work around it and move it back to common.h

alwa-nordic replied:
Possible solutions:
 - Make the variable const and add header guards. (I prefer this.)
 - Declare the variable extern in the header. Define the variable in common.c.
 - Wrap the variable in a static inline function. The variable must be static const.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_syncer.c:
```diff
@@ -22,12 +22,35 @@ extern enum bst_result_t bst_result;
+	0xEE, 0xEF, 0xF0, 0xF1, 0xF2, 0xF3, 0xF4, 0xF5, 0xF6, 0xF7, 0xF8, 0xF9, 0xFA, 0xFB};
+
+
+static const struct bt_data ad[] = {BT_DATA(BT_DATA_MANUFACTURER_DATA, mfg_data, 254),
+				    BT_DATA(BT_DATA_MANUFACTURER_DATA, mfg_data, 254)};
```
We only need the `mfg_data` and not the `bt_data` to compare, right?

Pieloaf replied:
oh actually yeah that would make more sense.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_syncer.c:
```diff
@@ -134,9 +157,43 @@ static void term_cb(struct bt_le_per_adv_sync *sync,
+		return;
+	}
+
+	char le_addr[BT_ADDR_LE_STR_LEN];
+	uint8_t buf_data_len;
```
Declare at top of scope please

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_syncer.c:
```diff
@@ -134,9 +157,43 @@ static void term_cb(struct bt_le_per_adv_sync *sync,
+	printk("PER_ADV_SYNC[%u]: [DEVICE]: %s advertisment received\n",
+	       bt_le_per_adv_sync_get_index(recv_sync), le_addr);
+
+	buf_data_len = (uint8_t)net_buf_simple_pull_le16(buf);
+
+	if (buf->data[0]-1 != ad[0].data_len ||
+		memcmp(buf->data, ad[0].data, ad[0].data_len)) {
+		FAIL("Unexpected adv data received\n");
+	}
+
+	net_buf_simple_pull(buf, ad[0].data_len);
+	buf_data_len = (uint8_t)net_buf_simple_pull_le16(buf);
+
+	if (buf->data[0]-1 != ad[1].data_len ||
+		memcmp(buf->data, ad[1].data, ad[1].data_len)) {
+		FAIL("Unexpected adv data received\n");
+	}
```
Suggest to use `bt_data_parse` as that will also validate the length checks, and make it easier to perform the comparison for each LTV structure

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_syncer.c:
```diff
@@ -280,6 +337,29 @@ static void main_per_adv_conn_privacy_syncer(void)
+	printk("Waiting for periodic sync lost...\n");
+	WAIT_FOR_FLAG(flag_per_adv_sync_lost);
+#endif
+	PASS("Periodic advertising syncer passed\n");
```
```suggestion
	PASS("Periodic advertising long data syncer passed\n");
```

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_advertiser.c:
```diff
@@ -276,6 +313,31 @@ static void main_per_adv_conn_privacy_advertiser(void)
+	delete_adv_set(per_adv);
+	per_adv = NULL;
+#endif
+	PASS("Periodic advertiser passed\n");
```
```suggestion
	PASS("Periodic long data advertiser passed\n");
```

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_syncer.c:
```diff
@@ -307,6 +387,15 @@ static const struct bst_test_instance per_adv_syncer[] = {
 	},
+	{
+		.test_id = "per_adv_long_data_syncer",
+		.test_descr = "Periodic advertising sync test with maximum "
+			      "data length. Test is used to verify that "
```
What does `maximum data length` refer to here? We are not sending 1650 octets, so I guess it refers to the maximum data in a single LTV (bt_data) struct? If so, that can be done with just a single bt_data struct, but you are using two, so the description here doesn't seem to match the actual test case implementation

Thalley reviewed (changes requested):
Generally the implementation changes look good, and it's nice with the test case. Had a few comments for some improvements

Thalley commented:
@Pieloaf the commit "header" of the second commit should also be `tests: bluetooth` instead of `bluetooth: host`

Pieloaf committed:
bluetooth: host: drop incomplete data in periodic scanner

This changes the periodic scanner to drop periodic adv
reports with incomplete data. This avoids incorrect data
being sent to application in the case where the periodic
adv data is not successfully received by the scanner.

Fixes zephyrproject-rtos#54072

Signed-off-by: Pierce Lowe <pierce.lowe@nordicsemi.no>

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_advertiser.c:
```diff
@@ -145,6 +145,25 @@ static void start_per_adv_set(struct bt_le_ext_adv *adv)
+{
+	int err;
+	const struct bt_data ad[] = {
+		BT_DATA(BT_DATA_MANUFACTURER_DATA, mfg_data, 254),
+		BT_DATA(BT_DATA_MANUFACTURER_DATA, mfg_data, 254)};
```
Is there any issue with using `BT_DATA_BYTES` on an `extern`? Otherwise I suggest to do
```suggestion
		BT_DATA_BYTES(BT_DATA_MANUFACTURER_DATA, mfg_data),
		BT_DATA_BYTES(BT_DATA_MANUFACTURER_DATA, mfg_data)};
```

If that is not possible, I suppose `sizeof(mfg_data)` also won't work instead of hardcoding the size to `254`

Pieloaf replied:
Sorry I replied to one of the older comments regarding this but yes: 
I was getting build errors with BT_DATA_BYTES as it was expecting a char/uint8_t not a uint8_t *
Looking at other instances where BT_DATA_BYTES is used the data bytes are inline in the function and not a pointer to an array of the data. So have left it as BT_DATA 

But yeah using sizeof(mfg_data) sounds good

Pieloaf replied:
should this also use ARRAY_SIZE instead of sizeof 
re: your other comment?

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_per_adv/src/per_adv_syncer.c:
```diff
@@ -134,9 +135,37 @@ static void term_cb(struct bt_le_per_adv_sync *sync,
+
+	while (buf->len > 0) {
+		buf_data_len = (uint8_t)net_buf_simple_pull_le16(buf);
+		if (buf->data[0]-1 != ARRAY_SIZE(mfg_data) ||
```
```suggestion
		if (buf->data[0] - 1 != sizeof(mfg_data) ||
```

The length field is in octets, and not number of elements. Of course when using `uint8_t` that's the same, but in case that `mfg_data` would every change type from `uint8_t`, using `ARRAY_SIZE` wouldn't work

alwa-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -805,6 +805,14 @@ void bt_hci_le_per_adv_report(struct net_buf *buf)
 		} else {
+			if (evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_INCOMPLETE) {
+				LOG_WRN("Received incomplete advertising data. "
+					"Advertising report dropped.");
```
This condition is expected during normal operation due to noise in the environment, right? Then it should not be a warning.

Pieloaf replied:
should it be replaced with a LOG_DBG?

Pieloaf commented on subsys/bluetooth/host/scan.c:
```diff
@@ -805,6 +805,14 @@ void bt_hci_le_per_adv_report(struct net_buf *buf)
 		} else {
+			if (evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_INCOMPLETE) {
+				LOG_DBG("Received incomplete advertising data. "
+					"Advertising report dropped.");
```
I'm not sure exactly what it would be used for, it was something I considered when making the change, if the data should just be passed to the application as incomplete and reset the buffer or drop the data before resetting the before. 
I chose to drop the data as the same was already being done for the extended scanner: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/scan.c#L634-L640

jhedberg replied:
Ok. I have a feeling we might later have to add this back, possibly through an optional `Kconfig` option, in case an app for whatever reason wants to see even these packets.

jhedberg replied:
Anyway, I'm fine with making this consistent with the extended scanner for now.

sjanc replied:
I think the reason is to not have to cache on controller side,  you may have up to 7 chains with ~2.4s spacing for each of scanned device

Pieloaf replied:
I can create a separate issue/PR with a kconfig for this after its merged if that suits?

Thalley replied:
I remember we discussed this in detail when it was added. 

We definitely wanted to provide reassembled data to the upper layers, so they did not have to implement reassembly themselves. 

I don't fully recall the argument to drop partial reports, rather than providing them to the upper layers. Perhaps the idea was that we wanted to avoid providing an advertising report where the length field of the report, did not match the actual report, as that would also mean that something `bt_data_parse` would not work with the received report.
